File: 01-historian-hysteria/part-2/main.go
package main

import (
	"bufio"
	"log"
	"os"
	"strconv"
	"strings"
	"time"
)

func main() {
	startTime := time.Now()

	file, err := os.Open("./input")
	if err != nil {
		log.Fatalf("Error reading input: %v", err)
	}
	defer file.Close()

	numbersA := make([]int, 0)
	numbersAndCountB := make(map[int]int)

	scanner := bufio.NewScanner(file)

	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" {
			continue
		}

		tokens := strings.Fields(line)
		a, err := strconv.Atoi(tokens[0])
		if err != nil {
			log.Fatalf("Error converting item %s to int: %v", tokens[0], err)
		}

		b, err := strconv.Atoi(tokens[1])
		if err != nil {
			log.Fatalf("Error converting item %s to int: %v", tokens[0], err)
		}

		numbersA = append(numbersA, a)
		numbersAndCountB[b] += 1
	}

	similarity := 0
	for _, item := range numbersA {
		similarity += item * numbersAndCountB[item]
	}

	log.Printf("The result is %d\n. Took: %d microseconds", similarity, time.Since(startTime).Microseconds())
}

File: 01-historian-hysteria/part-2/main.py
left = []
numbersAndOccurrences = {}

with open("input", "r") as f:
    for line in f.readlines():
        tokens = line.split()

        t1, t2 = int(tokens[0]), int(tokens[1])

        left.append(t1)

        if t2 not in numbersAndOccurrences:
            numbersAndOccurrences[t2] = 0

        numbersAndOccurrences[t2] = numbersAndOccurrences[t2] + 1


similarity = 0

for item in left:
    if item in numbersAndOccurrences:
        similarity = similarity + (item * numbersAndOccurrences[item])

print(similarity)
